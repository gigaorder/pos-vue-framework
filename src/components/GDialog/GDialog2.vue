<template>
	<div ref="el" class="dialog2">
		<div ref="wrapper" class="dialog2-wrapper" v-if="lazyRender" :class="wrapperClasses" :style="wrapperStyles" :tabindex="tabIndex">
			<div class="dialog2-content" :class="contentClasses" :style="contentStyles" ref="content" v-click-outside:[directiveArgs]="directiveValue">
				<slot></slot>
			</div>
		</div>
		<g-overlay ref="overlay" class="dialog2-overlay" v-if="renderOverlay" v-model="isActive" :z-index="overlayZIndex"></g-overlay>
		<div ref="activator">
			<slot name="activator" :toggleOverlay="toggleDialog"></slot>
		</div>
	</div>
</template>

<script>
  import getVModel from '../../mixins/getVModel';
  import { getZIndex } from '../../utils/helpers';
  import detachable from '../../mixins/detachable';
  import stackable from '../../mixins/stackable';
  import { computed, createElement as h, onMounted, reactive, ref, watch } from '@vue/composition-api';
  import ClickOutside from '../../directives/click-outside/click-outside';
  import GOverlay from '../GOverlay/GOverlay';

  export default {
    name: 'GDialog2',
    components: { GOverlay },
    directives: {
      ClickOutside
    },
		props: {
			value: {
			  type: Boolean,
				default: false
			},

			// Sizing
      maxWidth: {
        type: [String, Number],
        default: 'none',
      },

      width: {
        type: [String, Number],
        default: 'auto',
      },

			// Styling
			persistent: Boolean,
      hideOverlay: Boolean,
      scrollable: Boolean,
			fullscreen: Boolean,

			// Lazy/Eager
			lazy: Boolean,
		},
		setup (props, context) {
      const { model: isActive } = getVModel(props, context);
      const { attachToRoot, attachToParent } = detachable(props, context);
      const { getMaxZIndex } = stackable(props, context);

      // Stacking
      const zIndex = computed(() => {
        return !isActive.value ? 6 : getMaxZIndex(context.refs.wrapper) + 2
			});
      const overlayZIndex = computed(() => zIndex.value - 1);


      // Show/hide overlay
			// TODO: convert to overlayable mixin
			const renderOverlay = computed(() => !props.hideOverlay && !props.fullscreen && lazyRender.value);



      // Lazy/Eager
      // TODO: convert to bootable mixin
      const isBooted = reactive({
        value: false
      });
      const lazyRender = computed(() => isBooted.value || !props.lazy);

      function initComponent() {
        context.refs.wrapper.addEventListener('keydown', onKeydown);
        if (renderOverlay.value) attachToRoot(context.refs.overlay.$el);
        attachToRoot(context.refs.wrapper);
        attachToParent();
			}

      onMounted(() => {
        if (props.lazy) return;
        initComponent();
			});

			watch(isActive, newVal => {
			  if (newVal) {
          if (props.lazy) {
            isBooted.value = true;
            context.root.$nextTick(() => {
              initComponent();
            })
          }
          context.root.$nextTick(() => {
            context.refs.wrapper.focus();
          });
				}
			})

      function toggleDialog() {
        isActive.value = !isActive.value;
			}


			// Dynamic Classes and Styles
			const contentClasses = computed(() => ({
				'dialog2-content__active': isActive.value,
				'dialog2-content__scrollable': props.scrollable,
				'dialog2-content__fullscreen': props.fullscreen
			}));

      const contentStyles = computed(() => ({
        maxWidth: props.maxWidth === 'none' || props.fullscreen ? undefined : props.maxWidth,
        width: props.width === 'auto' || props.fullscreen ? undefined : props.width,
			}));

      const wrapperClasses = computed(() => ({
				'dialog2-wrapper__active': isActive.value
			}));

      const wrapperStyles = computed(() => ({
				zIndex: zIndex.value
			}));


      // Click outside
      const closeConditional = (e) => {
        if (!isActive.value) return false;
        const clickedInsideContent = context.refs.content.contains(e.target);
				if (clickedInsideContent) return false;

				// If z-index of current element is lower than the current active z-index then do not close when click outside
        return getZIndex(context.refs.wrapper) >= getMaxZIndex(context.refs.wrapper);
      };
      const directiveValue = () => {
        if (!props.persistent) {
          isActive.value = false
        }
      };
			const directiveArgs = {
        closeConditional,
				include: () => []
      };

			// Press ESC key
			const tabIndex = computed(() => isActive.value ? 0 : undefined);

			function onKeydown(e) {
				if (e.keyCode === 27) {
          isActive.value = !isActive.value;
				}
        context.root.$emit('keydown', e);
			}

      return {
        isActive,
				renderOverlay,
				overlayZIndex,
				toggleDialog,
				contentClasses,
				contentStyles,
				wrapperClasses,
				wrapperStyles,
        directiveValue,
				directiveArgs,
				tabIndex,
				lazyRender
			}
		}
  }
</script>

<style scoped>

</style>