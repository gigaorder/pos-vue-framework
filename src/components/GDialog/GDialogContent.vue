<script>
  import { getInternalValue } from '../../mixins/getVModel'
  import { convertToUnit, getZIndex } from '../../utils/helpers'
  import detachable from '../../mixins/detachable'
  import stackable from '../../mixins/stackable'
  import { computed, onBeforeUnmount, onMounted, ref, watch } from '@vue/composition-api'
  import ClickOutside from '../../directives/click-outside/click-outside'
  import GOverlay from '../GOverlay/GOverlay'
  import dependent from '../../mixins/dependent';

  export default {
    name: 'GDialogContent',
    components: {GOverlay},
    directives: {
      ClickOutside
    },
    props: {
      value: {
        type: Boolean,
        default: false
      },

      // Dialog Sizing
      maxWidth: {
        type: [String, Number],
        default: 'none',
      },

      width: {
        type: [String, Number],
        default: 'auto',
      },

      // Dialog functionality
      persistent: Boolean,
      hideOverlay: Boolean,
      scrollable: Boolean,
      fullscreen: Boolean,

      // Overlay styling
      overlayColor: String,
      overlayOpacity: [Number, String],

      bottom: Boolean,

      //content class for styling
      contentClass: String,

      // dependent mixin
      ...{
        closeDependents: {
          type: Boolean,
          default: true,
        },
        isDependent: {
          type: Boolean,
          default: true,
        },
      },
    },
    setup(props, context) {
      const isActive = getInternalValue(props, context)
      const {attachToRoot, detach} = detachable(props, context)
      const {getMaxZIndex} = stackable(props, context)

      // Stacking
      const wrapperZIndex = computed(() => {
        return !isActive.value ? 6 : getMaxZIndex(context.refs.wrapper) + 2
      });
      const overlayZIndex = computed(() => wrapperZIndex.value - 1)

      // Show/hide overlay
      // TODO: convert to overlayable mixin
      const renderOverlay = computed(() => !props.hideOverlay && !props.fullscreen)
      const getOpenDependentElements = ref(null)
      function initComponent() {
        attachToRoot(context.refs.overlay.$el)
        attachToRoot(context.refs.wrapper)
        detach(context.refs.container)
      }

      onMounted(function () {
        initComponent()
        getOpenDependentElements.value = dependent(this)
      })

      // Close conditional for click outside directive
      const closeConditional = (e) => {
        if (!isActive.value) {
          return false;
        }
        if (!context.refs.content) return;
        const clickedInsideContent = context.refs.content.contains(e.target);
        if (clickedInsideContent) {
          return false;
        }
        if (props.persistent) {
          return false;
        }

        // If z-index of current element is lower than the current active z-index then do not close when click outside
        return getZIndex(context.refs.wrapper) >= getMaxZIndex(context.refs.wrapper);
      };

      // Set the wrapper div tabindex to 0 when active, to make wrapper div focusable
      const wrapperTabIndex = computed(() => isActive.value ? 0 : undefined);

      // Change active state when press ESC
      function onKeydown(e) {
        if (e.key === 'Escape') {
          if (props.persistent) {
            return;
          }
          isActive.value = !isActive.value;
          context.emit('keydown', e);
        }
      }

      // Scroll prevent
      const composedPath = function (e) {
        if (e.composedPath) return e.composedPath()

        const path = []
        let el = e.target

        while (el) {
          path.push(el)

          if (el.tagName === 'HTML') {
            path.push(document)
            path.push(window)

            return path
          }

          el = el.parentElement
        }
        return path
      }

      const hasScrollbar = function (el) {
        if (!el || el.nodeType !== Node.ELEMENT_NODE) return false
        const style = window.getComputedStyle(el)
        return ['auto', 'scroll'].includes(style.overflowY) && el.scrollHeight > el.clientHeight
      }

      const onEdge = function (el) {
        return el.scrollTop === 0 || el.scrollTop + el.clientHeight === el.scrollHeight
      }

      const shouldScroll = function (el, delta) {
        if (el.scrollTop === 0 && delta < 0) return true
        return el.scrollTop + el.clientHeight === el.scrollHeight && delta > 0
      }

      const isInside = function (el, parent) {
        if (el === parent) {
          return true
        } else if (el === null || el === document.body) {
          return false
        } else {
          return isInside(el.parentNode, parent)
        }
      }

      const checkPath = function (e) {
        const path = e.path || composedPath(e)
        let delta = e.deltaY

        if (e.type === 'keydown' && path[0] === document.body) {
          const dialog = context.refs.wrapper
          // getSelection returns null in firefox in some edge cases, can be ignored
          const selected = window.getSelection().anchorNode
          if (dialog && hasScrollbar(dialog) && isInside(selected, dialog)) {
            return shouldScroll(dialog, delta)
          }
          return true
        }

        for (let index = 0; index < path.length; index++) {
          const el = path[index]

          if (el === document) return true
          if (el === document.documentElement) return true
          if (el === context.refs.content) return true

          if (hasScrollbar(el)) {
            return shouldScroll(el, delta)
          }
        }

        return true
      }

      const onWheel = function (e) {
        if ((context.refs.overlay && e.target === context.refs.overlay.$el.firstChild) || checkPath(e)) e.preventDefault()
      }

      const checkPathTouch = function (e) {
        const path = e.path || composedPath(e)
        let delta = touchStartY - touchMoveY
        for (let index = 0; index < path.length; index++) {
          const el = path[index]

          if (el === document) return true
          if (el === document.documentElement) return true
          if (el === context.refs.content) return true

          if (hasScrollbar(el)) {
            if (!onEdge(el)) return false
            return shouldScroll(el, delta)
          }
        }

        return true
      }

      let touchStartY
      let touchMoveY
      let touchFlag = false
      let shouldScrollTouchMove = false


      const onTouchStart = function (e) {
        touchFlag = true
        touchStartY = e.touches[0].clientY
      }

      const onTouchMove = function (e) {
        if (touchFlag) {
          touchMoveY = e.touches[0].clientY
          shouldScrollTouchMove = checkPathTouch(e)
          touchFlag = false
        }
        if (((context.refs.overlay && e.target === context.refs.overlay.$el.firstChild) || shouldScrollTouchMove) && e.cancelable) {
          e.preventDefault()
        }
      }

      const disableOutsideScroll = () => {
        window.addEventListener('wheel', onWheel, {passive: false})
        window.addEventListener('touchstart', onTouchStart, {passive: false})
        window.addEventListener('touchmove', onTouchMove, {passive: false})
      }

      const enableOutsideScroll = () => {
        window.removeEventListener('wheel', onWheel)
        window.removeEventListener('touchstart', onTouchStart)
        window.removeEventListener('touchmove', onTouchMove)
      }

      const unwatch = watch(isActive, newVal => {
        if (newVal) {
          disableOutsideScroll()
          context.root.$nextTick(() => {
            context.refs.wrapper && context.refs.wrapper.focus()
          })
        } else {
          enableOutsideScroll()
        }
      })

      // Clean-up when destroy
      onBeforeUnmount(() => {
        unwatch()
        enableOutsideScroll()
        context.refs.wrapper && detach(context.refs.wrapper);
        context.refs.overlay && detach(context.refs.overlay.$el);
      });

      // Render functions
      function genContent() {
        const wrapperData = {
          ref: 'wrapper',
          staticClass: 'g-dialog-wrapper',
          class: {
            'g-dialog-wrapper__active': isActive.value,
            'g-dialog-wrapper__bottom': props.bottom,
          },
          style: {
            zIndex: wrapperZIndex.value
          },
          attrs: {
            tabindex: wrapperTabIndex.value
          },
          on: {
            keydown: onKeydown
          }
        }

        const contentData = {
          ref: 'content',
          staticClass: 'g-dialog-content',
          class: {
            'g-dialog-content__active': isActive.value,
            'g-dialog-content__scrollable': props.scrollable,
            'g-dialog-content__fullscreen': props.fullscreen,
            [props.contentClass]: !!props.contentClass,
          },
          style: {
            maxWidth: props.maxWidth === 'none' || props.fullscreen ? undefined : convertToUnit(props.maxWidth),
            width: props.width === 'auto' || props.fullscreen ? undefined : convertToUnit(props.width),
          },
          directives: [
            {
              name: 'clickOutside',
              value: () => {
                isActive.value = false
              },
              arg: {
                closeConditional,
                include: () => [...getOpenDependentElements.value()]
              }
            }
          ]
        }

        const transitionName = computed(() => props.bottom ? 'dialog-bottom-transition' : 'dialog-transition')

        return <div {...wrapperData}>
          <transition name={transitionName.value}>
            <div {...contentData} vShow={isActive.value}>
              {context.slots.default ? context.slots.default() : undefined}
            </div>
          </transition>
        </div>
      }

      function genOverlay() {
        const overlayData = {
          ref: 'overlay',
          props: {
            zIndex: overlayZIndex.value,
            color: props.overlayColor,
            opacity: props.overlayOpacity
          }
        }

        return <g-overlay vOn:input={e => isActive.value = e} value={isActive.value} {...overlayData}
                          vShow={renderOverlay.value}/>
      }

      function genDialog() {
        return <div ref="container">
          {genContent()}
          {genOverlay()}
        </div>
      }

      return {
        genDialog
      }
    },
    render() {
      return this.genDialog()
    }
  }
</script>
<style scoped lang="scss">
  .g-dialog {

    &-wrapper {
      display: flex;
      align-items: center;
      justify-content: center;
      position: fixed;
      pointer-events: none;
      width: 100%;
      height: 100%;
      top: 0;
      left: 0;
      z-index: 6;
      outline: none;
    }

    &-content {
      transition: .3s cubic-bezier(0.25, 0.8, 0.25, 1), z-index 1ms;
      width: 100%;
      display: flex;

      &:not(.g-dialog-content__fullscreen) {
        max-width: 90%;
        max-height: 90%;
      }

      > * {
        width: 100%
      }

      > ::v-deep.g-card {
        height: auto;
        overflow-y: auto;
        flex: 1 1 100%;

        > .g-card-title {
          font-size: 2em;
          font-weight: 500;
        }

        > .g-card-actions {
          justify-content: flex-end;
        }
      }
    }

    &-content__active {
      pointer-events: auto;
    }

    &-content__scrollable {
      > ::v-deep.g-card {
        display: flex;
        flex: 1 1 100%;
        flex-direction: column;
        max-height: 100%;
        max-width: 100%;

        > .g-card-title, .g-card-actions {
          flex: 0 0 auto
        }

        > .g-card-text {
          backface-visibility: hidden;
          flex: 1 1 auto;
          overflow-y: auto;
        }
      }
    }

    &-content__fullscreen {
      border-radius: 0;
      margin: 0;
      height: 100%;
      position: fixed;
      overflow-y: auto;
      top: 0;
      left: 0;
      right: 0;

      > ::v-deep .g-card {
        height: 100%;
        min-height: 100%;
        min-width: 100%;
        margin: 0 !important;
        padding: 0 !important;
      }
    }

    &-wrapper__bottom {
      align-items: flex-end;
    }
  }
</style>
