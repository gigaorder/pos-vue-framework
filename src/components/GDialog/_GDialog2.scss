.dialog2 {
  pointer-events: auto;
  overflow-y: hidden;

  &-wrapper {
    display: flex;
    align-items: center;
    justify-content: center;
    position: fixed;
    pointer-events: none;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    z-index: 6;
  }

  &-content {
    display: none;

  }

  &-content__active {
    display: flex;
    pointer-events: auto;
  }
}