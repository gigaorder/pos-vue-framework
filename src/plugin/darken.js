import {getHslColor} from "../utils/colors";

;(function (window) {
  'use strict';

  const Darken = window.Darken || {};

  const Effect = {
    //click
    elements: [],
    // Effect delay
    duration: 1500,

    show: function (e, element) {

      // Disable right click
      if (e.button === 2) {
        return false;
      }

      const el = element || this;
      Effect.elements.push(el)

      const background = el.style.backgroundColor,
          border = el.style.borderColor,
          shadow = el.style.boxShadow

      let {h, s, l, a} = getHslColor(background)
      if (h === 0 && s === 0) {
        let hsla
        if (border)
          hsla = getHslColor(border)
        else
          hsla = {h: 0, s: 0, l: 0, a: 0}
        h = hsla.h
        s = hsla.s
        l = hsla.l
        a = hsla.a
      }
      el.style.background = `hsla(${h}, ${s - 5}%, ${l - 15}%, ${a})`
      el.style.borderColor = `hsla(${h + 2}, ${s - 10}%, ${l - 15}%, ${a})`
      el.style.boxShadow = `1px 0px 5px hsla(${h + 2}, ${s - 10}%, ${l - 15}%, ${a})`;

      setTimeout(() => {
        el.style.backgroundColor = background
        el.style.borderColor = border
        el.style.boxShadow = shadow
        const index = Effect.elements.findIndex(element => element === el)
        Effect.elements.splice(index, 1)
      }, Effect.duration)
    },

  };

  const TouchHandler = {
    touches: 0,
    allowEvent: function (e) {
      let allow = true;

      if (e.type === 'touchstart') {
        TouchHandler.touches += 1;
      } else if (e.type === 'touchend' || e.type === 'touchcancel') {
        setTimeout(function () {
          if (TouchHandler.touches > 0) {
            TouchHandler.touches -= 1;
          }
        }, Effect.duration);
      } else if (e.type === 'mousedown' && TouchHandler.touches > 0) {
        allow = false;
      }

      return allow;
    },
    touchup: function (e) {
      TouchHandler.allowEvent(e);
    }
  };

  function getWavesEffectElement(e) {
    if (TouchHandler.allowEvent(e) === false) {
      return null;
    }

    let element = null;
    let target = e.target || e.srcElement;

    while (target.parentNode !== null) {
      if (!(target instanceof SVGElement) && target.className.indexOf('darken-effect') !== -1) {
        element = target;
        break;
      }
      target = target.parentNode;
    }
    return element;
  }

  function showEffect(e) {
    const element = getWavesEffectElement(e);

    if (element !== null && Effect.elements.findIndex(el => el === element) === -1) {
      Effect.show(e, element);
    }
  }

  Darken.displayEffect = function (options) {
    options = options || {};

    if ('duration' in options) {
      Effect.duration = options.duration;
    }


    if ('ontouchstart' in window) {
      document.body.addEventListener('touchstart', showEffect, false);
    }

    document.body.addEventListener('mousedown', showEffect, false);
  };

  Darken.attach = function (element) {
    if ('ontouchstart' in window) {
      element.addEventListener('touchstart', showEffect, false);
    }

    element.addEventListener('mousedown', showEffect, false);
  };

  window.Darken = Darken;

  document.addEventListener('DOMContentLoaded', function () {
    Darken.displayEffect();
  }, false);

})(window);
