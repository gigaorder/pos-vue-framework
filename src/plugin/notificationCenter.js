import Vue from 'vue';
import GNotification from '../components/GNotification/GNotification';

export default {
  install(Vue, options) {
    const notificationCenter = new Vue({
      ...GNotification,
      propsData: options ? parseOptions(options) : {}
    })
    notificationCenter.$mount(document.body.appendChild(document.createElement('div')));

    Vue.prototype.$notificationCenter = notificationCenter;
  }
}

function parseOptions(options) {
  const propsData = {}

  if(options.position) {
    const position = options.position.split('-');
    for (let i = 0; i < position.length; i++) {
      propsData[position[i]] = true
    }
  }

  if (options.width) propsData['width'] = options.width

  return propsData;
}