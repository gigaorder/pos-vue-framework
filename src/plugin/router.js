import Vue from 'vue';
import VueRouter from 'vue-router';
import KeyboardDemo from '@/view/KeyboardDemo';
import Groupable from '@/view/Groupable';
import MenuDemo from '@/view/MenuDemo';
import SliderDemo from '../view/SliderDemo';
import OverlayDemo from "../view/OverlayDemo";
import SelectIonControl from '../view/SelectIonControl';
import DialogDemo from '../view/DialogDemo';
import PulseDemo from '../view/PulseDemo';
import ToolTipDemo from '../view/ToolTipDemo';
import InputDemo from "../view/InputDemo";
import TransitionDemo from '../view/TransitionDemo';
import ListDemo from '../view/ListDemo';
import List2Demo from "../view/List2Demo";
import DndDialogDemo from '../view/DndDialogDemo';
import GChipGroupDemo from '../view/GChipGroupDemo';
import DatePicker from '../view/DatePickerDemo';
import ExpansionPanelDemo from '../view/ExpansionPanelDemo';
import Textarea from "../view/TextareaDemo";
import LoginView from '../view/LoginView';

const routes = [
  {path: '/login', name: 'Login', component: LoginView},
];

const router = new VueRouter({
  mode: 'history',
  base: '/',
  routes
});

export default router
