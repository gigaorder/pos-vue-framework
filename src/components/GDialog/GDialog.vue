<script>
  import { getInternalValue } from '../../mixins/getVModel';
  import detachable from '../../mixins/detachable';
  import { computed, reactive, watch, onMounted, onBeforeUnmount } from 'vue';
  import ClickOutside from '../../directives/click-outside/click-outside';
	import GDialogContent from './GDialogContent';
  import {getScopeIdRender} from "../../utils/helpers";

  export default {
    name: 'GDialog',
    components: { GDialogContent  },
    directives: {
      ClickOutside
    },
    props: {
      modelValue: {
        type: Boolean,
        default: false
      },

      // Dialog Sizing
      maxWidth: {
        type: [String, Number],
        default: 'none',
      },

      width: {
        type: [String, Number],
        default: 'auto',
      },

      // Dialog functionality
      persistent: Boolean,
      hideOverlay: Boolean,
      scrollable: Boolean,
      fullscreen: Boolean,

      // Lazy/Eager
      eager: Boolean,

      // Overlay styling
      overlayColor: String,
      overlayOpacity: [Number, String],

      bottom: Boolean,

      //content class for styling
      contentClass: String,
    },
    emits: ['update:modelValue'],
    setup(props, context) {
      const isActive = getInternalValue(props, context);
      const { attachToParent, detach } = detachable(props, context)

      // Lazy/Eager
      // TODO: convert to bootable mixin
      const isBooted = reactive({
        value: false
      });
      const renderContent = computed(() => isBooted.value || props.eager);

      onMounted(() => {
        attachToParent(context.refs.activator)
      });

      const unwatch = watch(isActive, newVal => {
        if (newVal) {
          if (!props.eager) {
            isBooted.value = true;
          }
        }
      })

      function toggleDialog() {
        isActive.value = !isActive.value;
      }

      // Clean-up when destroy
      onBeforeUnmount(() => {
        unwatch()
        context.refs.activator && detach(context.refs.activator)
      })

      // Render functions
			function genDialogContent() {
        const dialogContentData = {
          props: {
            ...Object.assign({}, props, {
              value: isActive.value
            })
          },
          'onUpdate:modelValue': (value) => {
              isActive.value = value
					}
				}

				return <g-dialog-content {...dialogContentData}>
          {context.slots.default ? context.slots.default() : undefined}
				</g-dialog-content>
			}

      function genActivator() {
        return <div ref="activator" class="g-dialog-activator">
          {context.slots.activator ? context.slots.activator({ toggleDialog }) : undefined}
        </div>
      }

      function genDialog() {
        return <div ref="el" class="g-dialog">
          {genActivator()}
					{renderContent.value && genDialogContent()}
        </div>
      }

      function open() {
        isActive.value = true;
      }

      function close() {
        isActive.value = false;
      }

      return {
        open,
        close,
        isActive,
        genDialog
      }
    },
    render() {
      const scopeIdRender = getScopeIdRender()
      return scopeIdRender(this.genDialog.bind(this))()
    }
  }
</script>
<style scoped lang="scss">
	.g-dialog {
		pointer-events: auto;
		overflow-y: hidden;
		display: inline;
		position: relative;
	}
</style>
