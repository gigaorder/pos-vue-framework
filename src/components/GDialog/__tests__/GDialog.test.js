import { createElement } from '../../../tests/setup'

const Vue = require('vue/dist/vue.common.js')
import plugin from '@vue/composition-api'

Vue.use(plugin)

// Import your components after all of above import
import GDialog from '../GDialog';

describe('Dialog', () => {
  afterEach(() => {
    // clear document body after each test to prevent content slots from
    // mounting on the first div with [data-app] attr
    document.body.innerHTML = ''
  })
  describe('Single Dialog', () => {

    function build(template) {
      return new Vue({
        template: template,
        components: { GDialog },
        data() {
          return {
            dialog: false
          }
        }
      }).$mount(createElement());
    }

    it('Should render default and activator slots', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog">
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);

      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$children[0].$scopedSlots.default).toBeTruthy();
      expect(vm.$children[0].$scopedSlots.activator).toBeTruthy();
    })

    it('Should attach wrapper and overlay to root', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog">
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$el.firstChild.className).toBe('dialog-wrapper');
      expect(vm.$el.firstChild.nextSibling.className).toBe('overlay');
    })

    it('Should not render overlay if hide-overlay is set', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog" hide-overlay>
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$el.querySelector('.overlay')).toBe(null);
    })

    it('Dialog content should have __scrollable class if scrollable is set', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog" scrollable>
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$el.firstChild.firstChild.className).toBe('dialog-content dialog-content__scrollable');
    })

    it('Dialog content should have __fullscreen class if fullscreen is set', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog" fullscreen>
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$el.firstChild.firstChild.className).toBe('dialog-content dialog-content__fullscreen');
    })

    it('Should activate on click', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog">
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);

      const button = vm.$el.querySelector('button');
      const overlay = vm.$el.querySelector('.overlay');
      button.click();
      expect(vm.dialog).toBe(true);
      vm.$nextTick(() => {
        expect(vm.$el.outerHTML).toMatchSnapshot();
        expect(vm.$el.firstChild.className).toBe('dialog-wrapper dialog-wrapper__active');
      })
    })

    it('Should close when press ESC', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog">
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      const button = vm.$el.querySelector('button');
      const wrapper = vm.$el.querySelector('.dialog-wrapper');
      const mouseClick = new MouseEvent('click');
      button.dispatchEvent(mouseClick);
      expect(vm.dialog).toBe(true);

      vm.$nextTick(() => {
        const keypress = new KeyboardEvent('keydown', { key: 'Escape' });
        wrapper.dispatchEvent(keypress);
        expect(vm.dialog).toBe(false);
        vm.$nextTick(() => {
          expect(vm.$el.outerHTML).toMatchSnapshot();
        });
      })
    })

    it('Should not render on mounted if lazy is set. Should render when activated', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog" lazy>
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);
      expect(vm.$el.outerHTML).toMatchSnapshot();
      expect(vm.$el.querySelector('.dialog-wrapper')).toBe(null);
      expect(vm.$el.querySelector('.overlay')).toBe(null);

      const button = vm.$el.querySelector('button');
      button.click();
      expect(vm.dialog).toBe(true);
      vm.$nextTick(() => {
        vm.$nextTick(() => {
          expect(vm.$el.outerHTML).toMatchSnapshot();
          expect(vm.$el.firstChild.className).toBe('dialog-wrapper dialog-wrapper__active');
          expect(vm.$el.firstChild.nextSibling.className).toBe('overlay overlay__active');
        })
      })
    })

    it('Should remove all when destroy', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog">
            <template v-slot:activator="{ toggleDialog }">
              <button @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
        </div>`

      const vm = build(template);

      const button = vm.$el.querySelector('button');
      button.click();
      expect(vm.dialog).toBe(true);
      vm.$nextTick(() => {
        vm.$destroy()
        expect(vm.$el.outerHTML).toMatchSnapshot();
        expect(vm.$el.querySelector('.dialog-wrapper')).toBe(null);
        expect(vm.$el.querySelector('.overlay')).toBe(null);
        expect(vm.$el.querySelector('button')).toBe(null);
      })
    })
  })
  describe('Nested Dialog', () => {
    it('Nested dialog should have higher zIndex than outer dialog', () => {
      const template = `
        <div data-app>
          <g-dialog v-model="dialog1">
            <template v-slot:activator="{ toggleDialog }">
              <button id="button1" @click="toggleDialog">Dialog</button>
            </template>
            <div>content</div>
          </g-dialog>
          <g-dialog v-model="dialog2">
            <template v-slot:activator="{ toggleDialog }">
              <button id="button2" @click="toggleDialog">Dialog</button>
            </template>
            <div>
              <button id="buttonNested" @click="dialog1 = true">Open Dialog 1</button>
            </div>
          </g-dialog>
        </div>`

      const vm = new Vue({
        template: template,
        components: { GDialog },
        data() {
          return {
            dialog1: false,
            dialog2: false
          }
        }
      }).$mount(createElement());

      const dialogAll = vm.$el.querySelectorAll('.dialog-wrapper');
      const dialog1 = dialogAll['1'];
      const dialog2 = dialogAll['0'];
      const button1 = vm.$el.querySelector('#button1');
      const button2 = vm.$el.querySelector('#button2');
      const buttonNested = vm.$el.querySelector('#buttonNested');

      button2.click();
      expect(vm.dialog2).toBe(true);
      vm.$nextTick(() => {
        buttonNested.click();
        expect(vm.dialog1).toBe(true);
        vm.$nextTick(() => {
          expect(vm.$el.outerHTML).toMatchSnapshot();

          // Dialog1 is the nested dialog, Dialog2 is the outer dialog
          expect(dialog1.style.zIndex > dialog2.style.zIndex).toBe(true);
        })
      })
    })
  })
})