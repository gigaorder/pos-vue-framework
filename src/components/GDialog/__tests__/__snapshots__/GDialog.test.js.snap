// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Single Dialog Should activate on click 1`] = `
<div data-app>
  <div class="dialog-wrapper dialog-wrapper__active"
       style="z-index: 202;"
       tabindex="0"
  >
    <div class="dialog-content dialog-content__active">
      <div>
        content
      </div>
    </div>
  </div>
  <div class="overlay overlay__active"
       style="z-index: 201;"
  >
    <div class="overlay-scrim"
         style="background-color: rgb(33, 33, 33); opacity: 0.46;"
    >
    </div>
    <div class="overlay-content">
    </div>
  </div>
  <div class="dialog">
  </div>
  <button>
    Dialog
  </button>
</div>
`;
